---
title: "howto-pathway_analysis_hipathia"
output: 
    html_document:
        code_folding: hide
        highlight: tango
        number_sections: false
        toc: true
        toc_float:
            collapsed: false
            smooth_scroll: true
        toc_depth: 3
        theme: flatly
---
```{r}
require(hipathia)
```
# Key remarks
- data normalization and scaling methods
- pathways database

# Data preprocessing
```{r}
## load gene expression matrix
data('brca')
brca
## transform other gene ID to Entrez IDs
trans_data <- translate_data(brca_data, "hsa")
str(trans_data)
## scale data
exp_data <- normalize_data(trans_data, by_quantiles = TRUE)
```

```{r}
par(mfrow=c(2,1))
boxplot(trans_data)
boxplot(exp_data)
```

# Pathway activation computation
## factors to pass the signal:
  1. is the protein present? $\rightarrow$ quantify the presence of a particular gene as a normalized value between 0 and 1.
  2. is the protein "activated" by its neighbour? $\rightarrow$ compute signal value passign through a node taking into account the level of expression of each gene inside the node and the intensity of the signal arriving to it.
  3. the **signal value of the pathway** is the signal value through the last node of the pathway.
  
## concepts
  - **subpathways**: DAGs with input node that passes the signal until output node(s). Since the signal may follow many different paths, we compute the intensity of the signal up to each output node os a pathway separately (differently than FBA, where the intensity would split proportionally to the flux that can go through).
  - **effector proteins** are output nodes, reponsible for performing the action that the signal is seeking. Then, the **effector subpathway** is the subgraph including any node  leading to the selected effector protein.
  - To analyze which of the possible paths comprised in the effector subpathway are responsible for the observed change, we generate **decomposed subpathways** for each single input node and any node in the path to reach the output node.
  - **node expression**: some nodes (genes) may have multiple isoforms of the protein or members of the same gene family; since each gene has its own level of expression, first *hipathia* summarizes this information into a score representing the expression of the node as a whole.
  - **signal transduction**
    0. Initialize the pathway signal, assuming an incoming signal value of 1 in the input nodes of the subpathway.
    1. For each node $n$ of the network, the signal value $S_n$ is propagated along the nodes:
    
    $$S_n = v_n \cdot (1- \prod_{s_i \in A_n} (1- s_i) \cdot \prod_{s_j \in I_n} (1-s_j))$$
    where, $A_n$ is the set of signals arriving to the current node from an activation edge, $I_n$ is the set of signals arriving to the current node from an inhibition edge, and $v_n$ is the normalized value of expression of the current node.
    2. We obtain the signal value for each node (sample).
    
## usage of `hipathia`
- input: 
  - matrix of gene expression (genes x samples)
  - pathways object
  - parameters
    - decompose $\rightarrow$ to use effector subpathways or decomposed subpathways; less or more detail on the signal transduction. RECOMENDED USAGE: checkout everything first, decompose using a subset of the pathways
- output: levels of activation of the subpathways
  - The genes which are needed by hipathia to compute the signal and are not present in the
provided matrix are added by the function, assigning to each sample the median of the matrix

```{r}
## load pathways
pathways = load_pathways(species = 'hsa') # 
## computing the signal
results <- hipathia(exp_data, pathways, decompose = FALSE, verbose=TRUE)
```

# Output Analysis
```{r}
results
```
```{r}
experiments(results)[[1]]
```
```{r}
experiments(results)[[2]]
```

- *paths*: level of activity of the signal in each subpathway
  - rownames: IDs of the subpathways
```{r}
path_vals <- get_paths_data(results, matrix = TRUE)
dim(path_vals) # if decomposed, it would increase to 8440 subpathways
path_vals[50:60,1:5]
```
```{r}
# transform subpath IDs to comprehensive subpath names
get_path_names(metaginfo = pathways, names = c('P-hsa04010-15','P-hsa04010-67'), maxchar = NULL)
```

- **function activation computation**
To summarize the output, each effector protein performs a particular function; we can compute an intensity value for each molecular function and for each sample, and summarize the values of all the subpathways ending in an effector protein related to a function.
Each protein has a function annotation in different databases
```{r}
uniprot_vals = quantify_terms(results, pathways, dbannot = "uniprot")
uniprot_vals
```

```{r}
quantify_terms(results, pathways, dbannot = "GO")
```

# Example of pathway/function activation analysis
```{r}
## compare differences in paths between conditions
sample_group <- brca_design[colnames(path_vals),"group"]
comp <- do_wilcoxon(path_vals, sample_group, g1 = "Tumor", g2 = "Normal")
comp
```
```{r}
## retrieve overall significant pathways
get_pathways_summary(comp, metaginfo = pathways, conf = 0.05)
```
```{r}
## PCA 
ranked_path_vals <- path_vals[order(comp$p.value, decreasing = FALSE),]
pca_model <- do_pca(ranked_path_vals[1:ncol(ranked_path_vals),]) # take the ranked most significant as the number of rows cannot exceed the number of columns
pca_model$loadings[1:10,1:5]
```

```{r}
pca_plot(fit = pca_model, group = sample_group)
multiple_pca_plot(fit = pca_model, group = sample_group, plot_variance = TRUE)
```

```{r}
heatmap_plot(data = path_vals, group = sample_group)
```

```{r}
heatmap_plot(uniprot_vals, group = sample_group, colors="hipathia",
variable_clust = TRUE)
```

```{r}
colors_de <- node_color_per_de(results, pathways, sample_group, "Tumor",
"Normal")
pathway_comparison_plot(comp, metaginfo = pathways, pathway = "hsa03320",
node_colors = colors_de)
```

Colored edges represent significant subpathways: blue are down-activated, red are up-activated.

# Creating a new Pathways object - create own graphs
- example SIF
```
N-hsa00-A activation N-hsa00-C
N-hsa00-B inhibition N-hsa00-C
N-hsa00-C activation N-hsa00-D E
N-hsa00-D E activation N-hsa00-F
```
- example ATT
```
ID label X Y color shape type label.cex label.color width height genesList tooltip
N-hsa00-A A 0 40 white rectangle gene 0.5 black 46 17 998 A
N-hsa00-B B 0 0 white rectangle gene 0.5 black 46 17 5530 B
N-hsa00-C C 50 20 white rectangle gene 0.5 black 46 17 1432,5880,842 C
N-hsa00-D E D E 100 20 white rectangle gene 0.5 black 46 17 5747,/,9047,5335 D E
N-hsa00-F F 150 20 white rectangle gene 0.5 black 46 17 572 F
```

- read custom graphs from .sif files with attributes .att:
```{r}
newmgi <- mgi_from_sif(system.file("tmp/SIF_ATT_example",
                        package = "hipathia"))
```
  - requirements:
    - each pathway must be saved in two different files
    - the SIF and ATT files should have the same name $\rightarrow$ *hsa00.sif* and *hsa00.att* for pathway with id *hsa00*.
    - functions are not included in this files but annotated with a file of annotations from genes to functions
    - also a file including readable names of the pathways int he same folder: `name(dot)pathways(underscore)(species)(dot)txt`.
    
  - SIF file
    - text file with tab-separated columns
    - Each row is an interaction in the pathway: `source node(tab)type of relation(tab)target node`
    - Types of relations: activation or inhibition
    - name of the nodes in this file will be stored as the IDs of the nodes
    - nodes IDs should be structured as: `N(dash)pathway ID(dash)node ID`
    - Simple and complex nodes:
      - simple: many genes, but only one is needed to perform the function of the node
        - IDs like *N-hsa00-A*
      - complex nodes: different simple nodes and represent protein complexes. EAch simple node within the complex represents one protein in the complex; this node requires the presence of all their simple nodes to perform its function.
        - IDs separated by a space, like *N-hsa00-D E*.
    - ATT file
      - 12 tab-separated columns
      - each row is one node (simple or complex)
      - columns
        - node ID as above
        - label: first name in EntrezID; complex nodes, juxtapose gene names with spaces
        - X: position of the node in the pathway
        - Y: position of the node in the pathway
        - color: default color of the noe
        - shape: shape of the node
        - type: type of the node: "gene" or "compound", comma-separated for complex nodes
        - label.cex: rescaling for plotting
        - label.color
        - width: node with for plotting
        - height: node height for plotting
        - genesList: list of genes included in each node:
          - simple nodes: entrez ID of the genes included, separated by commas $\rightarrow$ *1432,5880,842 for node N-hsa00-C*.
          - complex nodes: genes list of the simple nodes included separated by a forward slash and no spaces, and in the same order as in the node ID $\rightarrow$ *For instance, node N-hsa00-D E includes two simple nodes: D and E. Its genesList column is 5747,/,9047,5335, meaning that the gene included in node D is 5747, and the genes included in node E are 9047 and 5335.*
        - tooltip: tooltip to be shown in the pathway visulaization; HTML code may be included.
        
    - pathway names file
      - text file with two columns separated by tabulars
      - each row a pathway: `id of the pathway(tab)real name of the pathway`
      - only activation and inhibition interactions are allowed
      
      
# References
- [Documentation from Bioconductor](https://www.bioconductor.org/packages/release/bioc/vignettes/hipathia/inst/doc/hipathia-vignette.pdf)
- [hpAnnot](https://bioconductor.org/packages/devel/data/annotation/vignettes/hpAnnot/inst/doc/hpAnnot-vignette.pdf)







